### ISSUE TITLE
aarch64: bitops.c :: int_of_bits_of_int : z3 can not prove: ensures x == int_of_bits(fst(bits_of_int(x, n)), snd(bits_of_int(x, n)))

### ISSUE THREAD

**Issue by stephengaito** [verifast/verifast#304]
Now that I can *compile* verifast on aarch4, I reach the following error during the self verification stage:

```
export PATH="/common/compiler/verifast/src/../bin/:/usr/local/bin:$PATH"; export LD_LIBRARY_PATH=":$LD_LIBRARY_PATH"; ../bin/verifast -c -prover z3v4.5 -emit_vfmanifest ../bin/bitops.c
../bin/bitops.c
../bin/bitops.c(7,15-17): Cannot prove condition. (Cannot prove (= x (int_of_bits (|(int) | (fst (bits_of_int x n))) (snd (bits_of_int x n))).)
make: *** [GNUmakefile:377: ../bin/bitops.vfmanifest] Error 1
root@bad988c6c368:/common/compiler# z3 --version
Z3 version 4.8.5 - 64 bit
```

Since Z3 is distributed on arm64 (for OSX).... and it compiled on the raspberry pi with no problems, I am assuming that this is a subtle endian problem...

Any suggestions?

By the way, verifast's theory is very well exposed and self verified! Many thanks for making this tool available!

**Comment by btj** [comment]
It's hard for me to debug this without being able to reproduce it, not having access to Arm hardware myself, but my first guess would be that this is just a case of Z3's performance/ability to prove a theorem sometimes being highly dependent on various environmental factors, such as, perhaps, the endianness of the platform it's running on.

**Comment by stephengaito** [comment]
@btj Yes I understand you won't be able to reproduce this problem... I was more hoping that you could give me hints as to how to proceed (which you have made an attempt at doing above).

As you can see from issue #307, I have tried to export the theory being verified and attempted to verify it using z3 directly on the command line... alas this did not work.

However even more interesting, I note that the output on x86_64 is 4916 lines of theory, while the output on aarch64 is only 3731 lines of theory.... so there is definitely something missing on the aarch64 (attempt) at compiling verifast.

This brings up a number of questions:

1. How does the theory for a verification get "compiled" into something z3 can work with? Which files are "merged" together to produce the input to libz3.so or the z3v4.5+SMTlib output?
2. Is there a difference between what is "fed" to libz3.so versus what gets output to z3v4.5-SMTlib output? (if so why?)
3. How do you routinely de-couple and debug new verifast theories?

**Comment by btj** [comment]
That the output on aarch64 is smaller is not surprising, given that z3 produces different responses. Note that VeriFast's interaction with the SMT solver is interactive: a different response from the solver will cause symbolic execution to proceed differently and therefore cause different queries to be generated. In this case, symbolic execution stops earlier.

1. The theory is generated on the fly as part of symbolic execution. Many of the initial axioms are from "executing" `bin/prelude.h`.
2. There should not be.
3. I have done this only once: https://github.com/verifast/verifast/issues/169
